ansible_builder/introspect.py

```diff
@@ -125,14 +163,76 @@ def get_dependency(self, entry):
         return req_file
 
 
+def simple_combine(reqs):
+    """Given a dictionary of requirement lines keyed off collections,
+    return a list with the most basic of de-duplication logic,
+    and comments indicating the sources based off the collection keys
+    """
+    consolidated = []
+    fancy_lines = []
+    for collection, lines in reqs.items():
+        for line in lines:
+            if line_is_empty(line):
+                continue
+
+            base_line = line.split('#')[0].strip()
+            if base_line in consolidated:
+                i = consolidated.index(base_line)
+                fancy_lines[i] += ', {}'.format(collection)
+            else:
+                fancy_line = base_line + '  # from collection {}'.format(collection)
+                consolidated.append(base_line)
+                fancy_lines.append(fancy_line)
+
+    return fancy_lines
+
+
+def write_files(data, write_pip=None, write_bindep=None):
+    if write_pip and data.get('python'):
+        with open(write_pip, 'w') as f:
+            f.write('\n'.join(simple_combine(data.get('python')) + ['']))
+    if write_bindep and data.get('system'):
+        with open(write_bindep, 'w') as f:
+            f.write('\n'.join(simple_combine(data.get('system')) + ['']))
+
+
+def parse_introspect_output(stdout):
+    p = re.compile(
+        '{0}(?P<yaml_text>.+){1}'.format(begin_delimiter, end_delimiter),
```

Comment:
Added regex processing here.

Why:

Before, we called `docker run -rm <tag> introspect`. When we did that, the stdout constituted a full YAML document (usually).

Here, we changed `introspect` to be inside a `RUN` layer, as opposed to a command ran outside of the Containerfile. That means that its output is interlaced in with the rest of the build command output. Instead of running more unnecessary commands, I decided I would just use markers to extract it.

Reply:
I don't love this. What was the reasoning here?

Reply:
Actually, https://github.com/ansible/ansible-builder/issues/78 factored into my thinking.

That issue wasn't important enough to act on, but then several other changes started getting thrown into the mix where `introspect` and `bindep` get thrown into a series of multiple commands. It became easier to assume that the output will get "padded" with other text, because it's not a hard problem to solve.